pkg/assembler/grammar.go

```diff
@@ -0,0 +1,199 @@
+package assembler
+
+import (
+	"fmt"
+	"math"
+)
+
+// Grammar and AST
+
+type CasmProgram struct {
+	Instructions []Instruction `@@`
+}
+
+type Instruction struct {
+	Core      *CoreInstruction `@@`
+	ApPlusOne bool             `( "," @"ap" "+" "+" )? ";" |`
```

Comment:
`call` cannot be followed by `ap++` according to the whitepaper (page 35).

One idea to fix this: we can remove `CoreInstruction` and "inline" its fields in `Instruction`. Then, we can use the `( ... )` grouping operator to specify that `ap++` can only follow `AssertEq`, `Jnz`, `Jump`, `ApPlus`, and `Ret`.

Reply:
Yep, thanks missed that. `ApPlus` (ap += <smth>), cannot be followed by an `ap+=1` as well. Also I am going to assume that Call instructions include `Ret` as well.

I am not sure how to use the (...) to solve it, if you could expand on that.

Over the top of my head, the updated grammar would look like this. 

 * Instruction => ApPlusInstruction | NonApPlusInstruction
 
With:

* NonApPlusInstruction => (ApPlus | Call | Ret ) ;

And:

* ApPlusInstruction => (Assert | Jmp) (,ap++ | ;)

I don't think we can inline it all into instruction, without to much repetition or making it very hard to read.
Also, this was an attempt by me to solve the `ap+=1` dilema at a parsing level, but we can also solve it at runtime level, although I wouldn't prefer that since it makes no sense for the instruction

Reply:
> ApPlus (ap += smth), cannot be followed by an ap++ as well. 

The whitepaper says, "The only instruction that may not use `ap++` is the `call` instruction."

> I am not sure how to use the (...) to solve it, if you could expand on that.

Something along the lines of:

```go
type Instruction struct {
	AssertEq  *AssertEq `(   @@`
	Jnz       *Jnz      `  | @@`
	Jump      *Jump     `  | @@`
	Ret       bool      `  | @ret`
	ApPlus    *ApPlus   `  | @@`
	ApPlusOne bool      `  ( "," @"ap" "+" "+" ";")? )`
	Call      *Call     `| @@`
}
```

Reply:
Making ret bool makes sense, but I thinks make the code harder to understand.

> The whitepaper says, "The only instruction that may not use ap++ is the call instruction."

The whitepaper is not clear enough sometimes. `ApPlus` is not an instruction per se, in the sense that it does not have an opcode (neither does any of the jumps).

But I am telling that they cannot be used because there is a conflict in the flags that are set: 

* `ApPlus` sets a flag to one
* `ApPlusOne` sets the same flag to two

So clearly, an inconsitency, and they cannot be together. If you check the cairo playground or the python compiler code you'd realize that they error as well in this cases

Reply:
`ApPlusOne` cannot be used with either `Ret` or `Call` because they forced the `Ap` to certain fixed behavior. And with`ApPlus` is a colition.

Reply:
Ok, so implemented something similar to your suggestion, completely removing `CoreInstruction` and also `CoreInstructioner` in https://github.com/NethermindEth/cairo-vm-go/pull/44/commits/c90023d6d940a55d9360b974093eaf723377e0c3